File: apps/api/athera_api/schemas/workspace.py
"""عقود مساحة عمل البحث | Project workspace contracts (PUBRIVA)."""
from __future__ import annotations

import datetime as dt
import uuid

from pydantic import BaseModel, Field

from ..models.screening import EXCLUSION_REASON_CODES

# النمط يُشتقّ من السجل ولا يُكتب بجانبه: قائمتان تفترقان بأول إضافة، فيقبل
# العقد رمزًا يرفضه القيد — أو يرفض رمزًا تعرضه الشاشة، وهو أسوأ.
REASON_CODE_PATTERN = "^(" + "|".join(EXCLUSION_REASON_CODES) + ")$"


class ProjectCreateRequest(BaseModel):
    """إنشاء بحث — **بأقلّ ما يلزم**.

    والباحث يبدأ بفكرة لا باستمارة: العنوان وحده يكفي، وبقيّة الحقول تُملأ
    حين تُعرف. واشتراطُ منهجٍ ونوع دراسة قبل أن يبدأ يوقفه عند الباب.
    """

    title_ar: str = Field(min_length=3, max_length=500)
    # من أين يبدأ: فكرة · مستند · بيانات · مخطوطة قائمة · فارغ
    starting_from: str = Field(
        default="idea", pattern="^(idea|document|dataset|manuscript|empty)$")


class ProjectRenameRequest(BaseModel):
    title_ar: str = Field(min_length=3, max_length=500)


class ProjectSummary(BaseModel):
    id: uuid.UUID
    title_ar: str
    status: str
    created_at: dt.datetime
    archived_at: dt.datetime | None = None
    deleted_at: dt.datetime | None = None
    files: int = 0
    sources: int = 0
    verified_facts: int = 0
    manuscripts: int = 0


class BrainEntryView(BaseModel):
    key: str
    label: str
    state: str = Field(description="known | needs_review | missing | conflicting")
    value: str | None = None
    sources: int = 0


class NextAction(BaseModel):
    key: str
    label: str


class ProjectOverview(BaseModel):
    project: ProjectSummary
    brain: list[BrainEntryView]
    recommended_next: NextAction | None = None
    blockers: list[str] = Field(default_factory=list)
    note: str


class ProjectFileView(BaseModel):
    file_id: uuid.UUID
    filename: str
    content_type: str
    size_bytes: int
    added_at: dt.datetime
    state: str
    processing_status: str = "not_processed"
    thesis_id: uuid.UUID | None = None
    candidates: int = 0
    reviewed: int = 0


class ProjectSourceView(BaseModel):
    source_id: uuid.UUID
    title: str
    doi: str | None = None
    publication_year: int | None = None
    use_state: str = Field(description="included | saved_only | excluded")
    added_at: dt.datetime
    decided_at: dt.datetime | None = None
    # سببُ الاستبعاد يُعاد مع الصفّ: حالٌ بلا سببها تُعرض حكمًا بلا تعليل،
    # فيقرأ الباحث «مستبعَدة» ولا يعرف لماذا — وهو ما يمنعه الترحيل 0023.
    exclusion_reason_code: str | None = None
    reason_ar: str | None = None


class LinkRequest(BaseModel):
    """ربط أصلٍ من المكتبة العامة بهذا البحث — بلا نسخ."""

    asset_id: uuid.UUID


class SourceUseRequest(BaseModel):
    """قرارُ الفرز — **مسارٌ واحد لا مساران**.

    شاشةُ الفرز وقائمةُ مراجع البحث تكتبان هنا كلتاهما: حالُ الاستعمال حقٌّ
    واحد في `project_sources`، ومسارٌ ثانٍ يكتبها كان سيصنع حقيقتين تفترقان
    — واحدةٌ تشترط سبب الاستبعاد وأخرى لا تشترطه.

    **والاستبعاد وحده يلزمه سبب.** أما «مُدرَج» و«محفوظ فقط» فلا: اشتراطُ
    تبريرٍ لكل قرار يجعل الباحث يكتب أيّ شيء ليمضي، فيصير الحقل ضجيجًا
    يُلغي قيمة السبب حيث يلزم فعلًا.
    """

    use_state: str = Field(pattern="^(included|saved_only|excluded)$")
    # رمزٌ من قائمةٍ مغلقة — يُعدّ ويُقارن ويُكتب منه قسم المنهجية.
    reason_code: str | None = Field(default=None, pattern=REASON_CODE_PATTERN)
    # النصّ الحرّ يرافق «سبب آخر» ويكون اختياريًّا مع غيره.
    reason_ar: str | None = Field(default=None, max_length=1000)


class ImpactView(BaseModel):
    """ما يترتب على الإزالة — **قبل أن تقع**."""

    is_safe: bool
    breaks_approved_work: bool
    summary: str
    consequences: list[dict] = Field(default_factory=list)


__all__ = ["BrainEntryView", "ImpactView", "LinkRequest", "NextAction",
           "ProjectCreateRequest", "ProjectFileView", "ProjectOverview",
           "ProjectRenameRequest", "ProjectSourceView", "ProjectSummary",
           "SourceUseRequest"]

File: apps/web/src/lib/workspace.ts
/**
 * مساحة عمل البحث | Project workspace client (PUBRIVA).
 *
 * **البحث هو الشيء المركزي.** فهذه الدوال تقرأ البحث بعلاقاته — ملفاته
 * ومراجعه وما تعرفه المنصّة عنه — لا وحداتٍ متجاورة يربط بينها الباحث في
 * رأسه.
 */
import { apiFetch } from "./api";
import type { Locale } from "./i18n";

export interface ProjectSummary {
  id: string;
  title_ar: string;
  status: string;
  created_at: string;
  archived_at: string | null;
  deleted_at: string | null;
  files: number;
  sources: number;
  verified_facts: number;
  manuscripts: number;
}

export interface BrainEntry {
  key: string;
  label: string;
  state: "known" | "needs_review" | "missing" | "conflicting";
  value: string | null;
  sources: number;
}

export interface ProjectOverview {
  project: ProjectSummary;
  brain: BrainEntry[];
  recommended_next: { key: string; label: string } | null;
  blockers: string[];
  note: string;
}

export interface ProjectFile {
  file_id: string;
  filename: string;
  content_type: string;
  size_bytes: number;
  added_at: string;
  state: string;
  processing_status: string;
  thesis_id: string | null;
  candidates: number;
  reviewed: number;
}

export interface ProjectSource {
  source_id: string;
  title: string;
  doi: string | null;
  publication_year: number | null;
  use_state: "included" | "saved_only" | "excluded";
  added_at: string;
  decided_at: string | null;
  /**
   * سببُ الاستبعاد يرافق حاله دائمًا.
   *
   * وحالٌ بلا سببها تُعرض حكمًا بلا تعليل: يقرأ الباحث «مستبعَدة» ولا يعرف
   * لماذا، فيعيد قراءة الدراسة — أو يخترع لها سببًا من ذاكرته.
   */
  exclusion_reason_code: string | null;
  reason_ar: string | null;
}

export interface Impact {
  is_safe: boolean;
  breaks_approved_work: boolean;
  summary: string;
  consequences: Array<{
    kind: string;
    count: number;
    label: string;
    breaks_approved_work: boolean;
  }>;
}

const base = "/api/v1/workspace";

export const listProjects = (locale: Locale, trash = false) =>
  apiFetch<ProjectSummary[]>(`${base}/projects?trash=${trash}`, { locale });

export const createProject = (locale: Locale, title: string) =>
  apiFetch<ProjectSummary>(`${base}/projects`, {
    locale,
    method: "POST",
    body: JSON.stringify({ title_ar: title }),
  });

export const archiveProject = (locale: Locale, id: string) =>
  apiFetch<ProjectSummary>(`${base}/projects/${id}/archive`, { locale, method: "POST" });

export const trashProject = (locale: Locale, id: string) =>
  apiFetch<Impact>(`${base}/projects/${id}`, { locale, method: "DELETE" });

export const restoreProject = (locale: Locale, id: string) =>
  apiFetch<ProjectSummary>(`${base}/projects/${id}/restore`, { locale, method: "POST" });

export const projectOverview = (locale: Locale, id: string) =>
  apiFetch<ProjectOverview>(`${base}/projects/${id}/overview`, { locale });

export const projectFiles = (locale: Locale, id: string) =>
  apiFetch<ProjectFile[]>(`${base}/projects/${id}/files`, { locale });

export const linkFile = (locale: Locale, id: string, fileId: string) =>
  apiFetch<ProjectFile>(`${base}/projects/${id}/files`, {
    locale,
    method: "POST",
    body: JSON.stringify({ asset_id: fileId }),
  });

export const fileImpact = (locale: Locale, id: string, fileId: string) =>
  apiFetch<Impact>(`${base}/projects/${id}/files/${fileId}/impact`, { locale });

/**
 * الإزالة تُقرّ قبل أن تقع.
 *
 * فالخادم يرفض بـ409 ما دام يقطع سندًا عن عملٍ اعتمده الباحث، ولا يُمرَّر
 * `acknowledged` إلا بعد أن يرى ما يترتب ويوافق — **والتحذير بعد الفعل ليس
 * تحذيرًا**.
 */
export const unlinkFile = (
  locale: Locale,
  id: string,
  fileId: string,
  acknowledged = false,
) =>
  apiFetch<Impact>(
    `${base}/projects/${id}/files/${fileId}?acknowledged=${acknowledged}`,
    { locale, method: "DELETE" },
  );

export const projectSources = (locale: Locale, id: string) =>
  apiFetch<ProjectSource[]>(`${base}/projects/${id}/sources`, { locale });

/**
 * قرّر حال المرجع في هذا البحث.
 *
 * **والاستبعاد يحمل سببه معه.** فالخادم يردّ استبعادًا بلا رمزٍ من قائمته
 * المغلقة، و«سبب آخر» يلزمه نصّه. وإرسالُ الحال أولًا ثم السبب في طلبٍ ثانٍ
 * كان سيترك دراسةً مستبعَدة بلا سبب لو فشل الثاني — وهي الحال نفسها التي
 * أُنشئ الحقل ليمنعها.
 */
export const setSourceUse = (
  locale: Locale,
  id: string,
  sourceId: string,
  useState: ProjectSource["use_state"],
  reason?: { reason_code?: string; reason_ar?: string },
) =>
  apiFetch<ProjectSource>(`${base}/projects/${id}/sources/${sourceId}`, {
    locale,
    method: "PATCH",
    body: JSON.stringify({ use_state: useState, ...(reason ?? {}) }),
  });

/** مصدرٌ في مكتبة الباحث — مرشَّحٌ للربط ببحث. */
export interface LibrarySource {
  id: string;
  title: string;
  doi: string | null;
  publication_year: number | null;
  journal_name: string | null;
  authors: string[] | null;
}

export const listLibrarySources = (locale: Locale) =>
  apiFetch<LibrarySource[]>("/api/v1/sources", { locale });

/**
 * اربط مرجعًا من المكتبة بهذا البحث.
 *
 * **مسارٌ كان في الخادم بلا بابٍ في المتصفح.** `POST …/sources` قائمٌ
 * ويُنشئ العلاقة بحال `saved_only`، ولم تكن في الواجهة نافذةٌ تبلغه —
 * فالباحث يرى مراجعه ولا يستطيع أن ينسبها إلى بحثه.
 *
 * والحال الابتدائية تأتي من الخادم لا من هنا: **الاستيراد ليس حكمًا
 * بالصلاحية دليلًا**، والترقية إلى «مُدرَج» قرارُ الباحث وحده.
 */
export const linkSource = (locale: Locale, id: string, sourceId: string) =>
  apiFetch<ProjectSource>(`${base}/projects/${id}/sources`, {
    locale,
    method: "POST",
    body: JSON.stringify({ asset_id: sourceId }),
  });
